### PR TITLE
backupccl: preserve column names and column nullability in EXPORT PARQUET

### LINKED ISSUE #72356: backupccl: add column names to parquet export
**Is your feature request related to a problem? Please describe.**
Currently the parquet file(s) created by EXPORT PARQUET do not contain the column names from the exported relation

**Describe the solution you'd like**
By default, the column names should be added to the exported parquet file

Epic CRDB-10340

### LINKED ISSUE #72357: backupccl: support null values in EXPORT PARQUET
**Is your feature request related to a problem? Please describe.**
Currently EXPORT PARQUET fails if it attempts to export a relation with a column with a null value.

**Describe the solution you'd like**
EXPORT PARQUET should handle null values of any type currently supported by EXPORT PARQUET

Epic CRDB-10340

### PR DESCRIPTION
**Pull request by msbutler**
Previously, parquet files created with EXPORT didn't retain their column names.
This PR passes the column names into the exported parquet file.

Previously, EXPORT PARQUET considered all CRDB columns as nullable, including
CRDB columns with the NOT NULL property. In this change, EXPORT PARQUET
preserves the NOT NULL property. This change gets us closer to a roundtrippable
EXPORT/IMPORT PARQUET.

Informs: #67710
Fixes: #72356
Fixes: #72357

Release note (sql change): EXPORT PARQUET now preserves column names and
nullability.

### CODE REVIEW

**Comment by msbutler** [comment]
oops, i refactored something i didn't intend to. hold tight.

**Comment by msbutler** [comment]
ok, review away!

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-:-Mpx83d98DPWN9Nl4jwl:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @msbutler, and @stevendanna)

___
*[pkg/ccl/importccl/exportparquet.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-Mpx6hhiD441MsBR5TK0:-Mpx6hhiD441MsBR5TK1:b-ay5shw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff04034f149fb19d860c89ad89912176a065364/pkg/ccl/importccl/exportparquet.go#L294)):*
> ```Go
> 
> // MustBeStreaming currently never gets called by the parquetWriterProcessor as the function only
> //applies to implementation
> ```

nit: comments should be 80 columns wide
___
*[pkg/sql/distsql_physical_planner.go, line 3954 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-MpwVn1T4cEj7fgAKsIS:-MpwVn1T4cEj7fgAKsIT:b-y0onlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff04034f149fb19d860c89ad89912176a065364/pkg/sql/distsql_physical_planner.go#L3954)):*
> ```Go
> 			ChunkSize:       n.chunkSize,
> 			UserProto:       planCtx.planner.User().EncodeProto(),
> 			LogicalColNames: n.logicalColNames,
> ```

nit: why use the name "LogicalColNames"? Why not just "ColNames"?
___
*[pkg/sql/export.go, line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-MpwVXsf2MqzAryA8d0N:-MpwVXsf2MqzAryA8d0O:b7kfo1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff04034f149fb19d860c89ad89912176a065364/pkg/sql/export.go#L176)):*
> ```Go
> 		colNames[i] = col.Name
> 		nullableCols[i] = true
> 		if notNullColSet.Contains(i) {
> ```

nit: the default is false, so you could just set this to true `if !notNullColSet.Contains(i)`
___
*[pkg/sql/execinfrapb/processors_bulk_io.proto, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-Mpx8LAr-m0HDY5jKcdB:-Mpx8LAr-m0HDY5jKcdC:bdr5hqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff04034f149fb19d860c89ad89912176a065364/pkg/sql/execinfrapb/processors_bulk_io.proto#L305)):*
> ```protobuf
>   optional FileCompression compression_codec = 5 [(gogoproto.nullable) = false];
> 
>   // logical_col_names specifies the logical column names for the exported parquet file
> ```

nit: comment should end with a period.
___
*[pkg/sql/opt/exec/factory.opt, line 733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-Mpx6FXW8IyOWiltnhHN:-Mpx6FXX341sgqVexlFy:bfjjay9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff04034f149fb19d860c89ad89912176a065364/pkg/sql/opt/exec/factory.opt#L733)):*
> ```
>     FileFormat string
>     Options []exec.KVOption
>     NotNullColsSet exec.NodeColumnOrdinalSet
> ```

super nit: why not just call these `NotNullCols`?

**Review by stevendanna — COMMENTED** [review]
Overall it looks pretty reasonable to me. I left some very minor nits and a couple fo questions.

File: pkg/ccl/importccl/exportparquet.go [hunk 1]
```diff
@@ -290,6 +290,8 @@ func (sp *parquetWriterProcessor) OutputTypes() []*types.T {
 	return res
 }
 
+// MustBeStreaming currently never gets called by the parquetWriterProcessor as the function only
+//applies to implementation
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/importccl/exportparquet.go:294]: Done.

File: pkg/sql/export.go [hunk 2]
```diff
@@ -161,6 +166,18 @@ func (ef *execFactory) ConstructExport(
...
+
+	cols := planColumns(input.(planNode))
+	colNames := make([]string, len(cols))
+	nullableCols := make([]bool, len(cols))
+	for i, col := range cols {
+		colNames[i] = col.Name
+		nullableCols[i] = true
+		if notNullColSet.Contains(i) {
+			nullableCols[i] = false
+		}
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/sql/export.go:178]: Done.

File: pkg/ccl/importccl/exportcsv_test.go [hunk 3]
```diff
@@ -268,12 +269,26 @@ func TestExportOrder(t *testing.T) {
...
+// If a validation field is empty, then that field will not be used in the test.
+type parquetTest struct {
+	name string
+	stmt string
+
+	// colNames provides the expected column names for the parquet file.
+	colNames []string
+
+	// nullableCols provides the expected nullability of each column in the parquet file.
+	nullableCols []parquet.FieldRepetitionType
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/importccl/exportcsv_test.go:283]: Done.

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 4]
```diff
@@ -302,9 +302,15 @@ message ParquetWriterSpec {
   // compression_codec specifies compression used for exported file. THIS IS ISN'T USED RIGHT NOW
   optional FileCompression compression_codec = 5 [(gogoproto.nullable) = false];
 
+  // logical_col_names specifies the logical column names for the exported parquet file
+  repeated string logical_col_names = 6 [(gogoproto.nullable) = false];
+
+  // nullable_cols specifies which columns allow null values in the exported parquet file
+  repeated bool nullable_cols = 9 [(gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:309]: Done. Going with col_names and col_nullability.

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 5]
```diff
@@ -302,9 +302,15 @@ message ParquetWriterSpec {
...
+
   // User who initiated the export. This is used to check access privileges
   // when using FileTable ExternalStorage.
-  optional string user_proto = 6 [(gogoproto.nullable) = false, (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/security.SQLUsernameProto"];
+  optional string user_proto = 8 [(gogoproto.nullable) = false, (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/security.SQLUsernameProto"];
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:313]: Don't know why i changed that. Changing it Back. Done.

**Review by rytaft — APPROVED** [review]
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @msbutler, and @stevendanna)

___
*[pkg/ccl/importccl/exportparquet.go, line 294 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-MqFRUmQ7tqG3eOgRkew:-MqFRUmQ7tqG3eOgRkex:b-qorz7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92e37f276fc65b4761b5610df03203265809e42/pkg/ccl/importccl/exportparquet.go#L294)):*
> ```Go
> 
> // MustBeStreaming currently never gets called by the parquetWriterProcessor as
> // the function only applies to implementation
> ```

nit: end comments with a period.
___
*[pkg/sql/export.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-MqFRdst9egSOws5UNd6:-MqFRdst9egSOws5UNd7:b1hhs82) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92e37f276fc65b4761b5610df03203265809e42/pkg/sql/export.go#L243)):*
> ```Go
> 		fileCompression: codec,
> 		colNames:        colNames,
> 		colNullability:  nullableCols,
> ```

nit: maybe rename `nullableCols` to `ColNullability` to match

**Review by stevendanna — APPROVED** [review]

**Review by rytaft — APPROVED** [review]
Sorry for the delayed comment, but do you know if you need to bump the DistSQL version for changes to processors_bulk_io.proto? https://github.com/cockroachdb/cockroach/blob/745d95eb5c7c104e2fedeed33c69fb197301b57d/pkg/sql/execinfra/version.go#L42

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-:-MqRkv0S4tFLXQx790TL:b-f242ep)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @stevendanna)

**Review by msbutler — COMMENTED** [review]
I don't think I need to. When I've made changes to this proto in the past, i haven't bumped the DistSQL version. But I don't know much about DistSQL versioning. What's the context of your question? Do some teams always update the DistSQLVersion whenever they make changes to a proto object that informs a processor?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-:-MqW431R9JvtUteRfh9S:bbvsvp6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @stevendanna)

___
*[pkg/sql/export.go, line 243 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-MqFRdst9egSOws5UNd6:-MqW5HDQEJgGcy7MSgyi:b-896fjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92e37f276fc65b4761b5610df03203265809e42/pkg/sql/export.go#L243)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: maybe rename `nullableCols` to `ColNullability` to match
</blockquote></details>

Done

**Review by rytaft — COMMENTED** [review]
The version is used to determine whether it's safe to distribute an operation and which nodes can be involved. I think you just have to think carefully about what happens if not all nodes that are involved in the DistSQL flow are aware of these two new fields. Generally it's safer to err on the side of bumping the version if you update any of the protos used by DistSQL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-:-MqWRUmP4HjUQpYaPLvb:bfnjhyo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @stevendanna)

**Review by msbutler — COMMENTED** [review]
I see. Given that EXPORT Parquet isn't on any release branches yet, and that I plan to make future changes to this proto, I think I'll wait to bump DistSQL's version. Thank you for the review!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73382#-:-MqWdE52Eo8EoH2LKfjR:b-4oqn2a)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @stevendanna)

**Comment by msbutler** [comment]
bors, r=rytaft, stevendanna

**Comment by msbutler** [comment]
bors, r=rytaft, stevendanna

**Comment by msbutler** [comment]
bors r=rytaft, stevendanna

**Comment by RaduBerinde** [comment]
Hi,
I noticed these warnings:
```
WARNING: field ParquetWriterSpec.ColNames is a repeated non-nullable native type, nullable=false has no effect
WARNING: field ParquetWriterSpec.ColNullability is a repeated non-nullable native type, nullable=false has no effect
```

We should remove the `nullable` thing for the repeated fields.

**Comment by msbutler** [comment]
Fixing this in #73739

**Pull request merged**